src/mlpack/tests/main_tests/softmax_regression_test.cpp

```diff
@@ -0,0 +1,289 @@
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+  SetInputParam("labels", std::move(labels));
+  SetInputParam("max_iterations", (int) -1);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Ensure that lambda is always non-negative.
+ */
+BOOST_AUTO_TEST_CASE(SoftmaxRegressionLambdaTest)
+{
+  arma::mat inputData;
+  if (!data::Load("trainSet.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset trainSet.csv!");
+
+  // Get the labels out.
+  arma::Row<size_t> labels(inputData.n_cols);
+  for (size_t i = 0; i < inputData.n_cols; ++i)
+    labels[i] = inputData(inputData.n_rows - 1, i);
+
+  // Delete the last row containing labels from input dataset.
+  inputData.shed_row(inputData.n_rows - 1);
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+  SetInputParam("labels", std::move(labels));
+  SetInputParam("lambda", (double) -0.1);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Ensure that number_of_classes is always positive.
+ */
+BOOST_AUTO_TEST_CASE(SoftmaxRegressionNumClassesTest)
+{
+  arma::mat inputData;
+  if (!data::Load("trainSet.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset trainSet.csv!");
+
+  // Get the labels out.
+  arma::Row<size_t> labels(inputData.n_cols);
+  for (size_t i = 0; i < inputData.n_cols; ++i)
+    labels[i] = inputData(inputData.n_rows - 1, i);
+
+  // Delete the last row containing labels from input dataset.
+  inputData.shed_row(inputData.n_rows - 1);
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+  SetInputParam("labels", std::move(labels));
+  SetInputParam("number_of_classes", (int) -1);
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Make sure only one of training data or pre-trained model is passed.
+ */
+BOOST_AUTO_TEST_CASE(SoftmaxRegressionTrainingVerTest)
+{
+  arma::mat inputData;
+  if (!data::Load("trainSet.csv", inputData))
+    BOOST_FAIL("Cannot load train dataset trainSet.csv!");
+
+  // Get the labels out.
+  arma::Row<size_t> labels(inputData.n_cols);
+  for (size_t i = 0; i < inputData.n_cols; ++i)
+    labels[i] = inputData(inputData.n_rows - 1, i);
+
+  // Delete the last row containing labels from input dataset.
+  inputData.shed_row(inputData.n_rows - 1);
+
+  // Input training data.
+  SetInputParam("training", std::move(inputData));
+  SetInputParam("labels", std::move(labels));
+
+  mlpackMain();
+
+  // Input pre-trained model.
+  SetInputParam("input_model",
+                std::move(CLI::GetParam<SoftmaxRegression>("output_model")));
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
```

Comment:
Looks good, but I have some simple ideas for a couple more tests that can help ensure that the values of lambda, no_intercept, and max_iterations are actually being used:

 * Build two models with different lambda values... ensure that the `SoftmaxRegression` objects returned have different parameters.

 * Build two models with different maximum numbers of iterations... ensure that the `SoftmaxRegression` objects returned have different parameters.

 * Ensure that a `SoftmaxRegression` model with no intercept term has one fewer parameter than one with an intercept term.

Like I said with the random forest tests, these are just to make sure that the parameters are actually properly being passed to the algorithm.  If you'd like to add these tests (or similar tests with the same goal), that would be great; and if not, that's no problem at all, we can make a note of it in the issue.

Reply:
Added all the above suggested tests. 
Not able to use checkmatrices() here for checking different parameters as log from BOOST_REQUIRE_EQUAL() wasn't getting suppressed.
Earlier thought of something like this:
> Log::Fatal.ignoreInput = true;
>  BOOST_REQUIRE_THROW(CheckMatrices(modelParam,
>        CLI::GetParam<SoftmaxRegression>("output_model").Parameters()),
>        std::exception);
> Log::Fatal.ignoreInput = false;